File: tmpevents/2021/02/02.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<(i32, i32)> {
    input
    .lines()
    .map(|command| command
        .splitn(2,' ')
        .collect_tuple()
        .unwrap())
    .map(|(dir, steps)| match dir {
        "forward"   => (steps.parse().unwrap(), 0),
        "down"      => (0, steps.parse().unwrap()),
        "up"        => (0, -steps.parse::<i32>().unwrap()),
        _ => unreachable!()})
    .collect_vec()   
}

fn part_1(commands: Vec<(i32, i32)>) -> i32 {
    solver(commands, false)
}

fn part_2(commands: Vec<(i32, i32)>) -> i32 {
    solver(commands, true)
}

fn solver(commands: Vec<(i32, i32)>, track_aim: bool) -> i32 {
    let (a, b, _) = commands
    .iter()
    .fold((0,0,0), |(x, y, aim), (dx, dy)| {
    if track_aim {(x + dx, y + dx * aim, aim + dy)}
    else {(x + dx, y + dy, 0)}});
    a * b
}

File: events/2021/02/02.py
def preprocessing(puzzle_input):
    commands = []
    for command in puzzle_input.splitlines():
        dir, steps = command.split(' ')
        match dir:
            case 'forward': commands.append((int(steps), 0          ))
            case 'up':      commands.append((0         , -int(steps)))
            case 'down':    commands.append((0         , int(steps) ))
    return commands

def solver(commands):
    x, y, y_aim, aim = 0, 0, 0, 0
    
    for dx, dy in commands:
        x     += dx
        y     += dy
        aim   += dy
        y_aim += dx * aim
        
    yield x * y
    yield x * y_aim